### PR TITLE
Request evidence

### PR DESCRIPTION
**Pull request by arturkow2000**

### CODE REVIEW

File: src/client/proto.rs [hunk 1]
```diff
@@ -196,22 +196,168 @@ where
...
+    Sha256,
+    Sha384,
+    Sha512,
+    Unknown(u32),
+}
+
+impl From<u32> for PcrAlgo {
+    fn from(x: u32) -> Self {
+        // FIXME: due to https://github.com/rust-lang/rust/issues/60553 we map
+        // identity map raw values into corresponding enum constants. Rust if free
```
Comment (krystian-hebel) [inline src/client/proto.rs]:
```suggestion
        // identity map raw values into corresponding enum constants. Rust is free
```

File: src/client/proto.rs [hunk 2]
```diff
@@ -196,22 +196,168 @@ where
     }
 }
 
-#[derive(Debug, Deserialize, Default)]
-pub struct Pcrs<'a> {
+#[derive(Debug, Copy, Clone)]
+#[repr(u32)]
```
Comment (krystian-hebel) [inline src/client/proto.rs]: Shouldn't `u16` be used instead?

File: src/client/proto.rs [hunk 3]
```diff
@@ -196,22 +196,168 @@ where
...
+}
+
+impl fmt::Display for PcrAlgo {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Sha1 => write!(f, "sha1"),
+            Self::Sha256 => write!(f, "sha256"),
+            Self::Sha384 => write!(f, "sha384"),
+            Self::Sha512 => write!(f, "sha512"),
+            Self::Unknown(u) => write!(f, "unknown (0x{:x})", u),
```
Comment (krystian-hebel) [inline src/client/proto.rs]:
```suggestion
            Self::Unknown(u) => write!(f, "unknown (0x{:04x})", u),
```
or `"unknown ({:#06x})"`. Full length u16 should be more readable than shortened version.

File: src/client/proto.rs [hunk 4]
```diff
@@ -233,3 +379,105 @@ impl<'a> Nonce<'a> {
...
+    }
+
+    fn do_verify_pcrs(pcrs: &PcrBank) -> Result<(), ()> {
+        // pcrs is a bitmask representing which PCRs are present and which are
+        // not.
+        let n1 = pcrs.pcrs.count_ones() as usize;
+        let n2 = pcrs.pcr.len();
+        if n1 != n2 {
+            error!(
+                "PCR count does not match, count from mask is {}, but really there are {} PCRs",
```
Comment (krystian-hebel) [inline src/client/proto.rs:398]: This doesn't get printed even if counts don't match. Depending on number of bits in `pcrs` next error about expected PCR size may be printed instead (if number of PCR values in not a multiply of bits set). Seems that both `n1` and `n2` are the same as number of bits in `pcrs`.

**Review by krystian-hebel — CHANGES_REQUESTED** [review]

File: src/client/proto.rs (hunk 2 above)
Comment (arturkow2000) [reply src/client/proto.rs]: Right, fixed

File: src/client/proto.rs (hunk 4 above)
Comment (krystian-hebel) [reply src/client/proto.rs:398]: Ok, it was my error during testing. This works as expected.

**Review by krystian-hebel — APPROVED** [review]

**Pull request merged**